### ISSUE TITLE
Can't send downlinks after the first one - Scheduler Device lock

### ISSUE THREAD

**Issue by joselunatic** [chirpstack/chirpstack#324]
I’m currently facing a challenge with downlink message handling in my ChirpStack environment. I am using ChirpStack version 4 with a Mikrotik ltAp gateway. The issue involves downlink messages getting stalled in the queue, and I’m encountering specific errors that I’m hoping to get help with.

Environment:

ChirpStack version: v4
Gateway model: Mikrotik ltAp
UDP packet forwarder
Issue Description : New devices can join the network and receive their first downlink message successfully. However, subsequent downlink messages get queued and are not sent. The chirpstack-gateway-bridge logs and chirpstack logs reveal two distinct error messages:

`nov 23 09:54:46 chirpstack-gateway-bridge[xxxxxx]: time="2023-11-23T09:54:46.690713473+01:00" level=error msg="backend/semtechudp: could not handle packet" addr="[IP Address]:[Port]" data_base64=Aj4IBUg2NyATAE8A error="no internal frame cache for token 7848"`

`ERROR chirpstack::downlink::scheduler: Schedule next queue-item for device failed error=Get device lock`

This behaviour repeats with every device. After the join, it can receive a downlink (after an uplink), but it can't send any further downlinks and the downlink queue never clears.

I've flushed all device queues, but the issue remains.

**Comment by joselunatic** [comment]
After adding gateways from a different manufacturer (Milesight UG65) I have noticed this behavior only happens with the Mikrotik gateways. Two Mikrotik on different geographical locations present the same issue (with different end-devices) while the Milesight gateways work properly.

**Comment by brocaar** [comment]
Could you please post the tcpdump of a downlink (example: https://www.chirpstack.io/docs/guides/connect-gateway.html#tcpdump)?

I'm interested in seeing the following messages:

* `PULL_RESP`
* `TX_ACK`

(https://github.com/Lora-net/packet_forwarder/blob/master/PROTOCOL.TXT#L274)

Based on the log:

```
nov 23 09:54:46 chirpstack-gateway-bridge[xxxxxx]: time="2023-11-23T09:54:46.690713473+01:00" level=error msg="backend/semtechudp: could not handle packet" addr="[IP Address]:[Port]" data_base64=Aj4IBUg2NyATAE8A error="no internal frame cache for token 7848"
```

It looks like the Mikrotik gateway is not using the correct token in the `TX_ACK` response (https://github.com/Lora-net/packet_forwarder/blob/master/PROTOCOL.TXT#L333).

**Comment by joselunatic** [comment]
Of course:
```
{"txpk":{"imme":false,"rfch":0,"powe":14,"ant":0,"brd":0,"tmst":3723346564,"freq":867.3,"modu":"LORA","datr":"SF10BW125","codr":"4/5","ipol":true,"size":32,"data":"YFjXewagAAAApKrOjDPKzYFkmTTuxrKlE9oVBmnOuWc="}}
12:15:20.067386 IP <host>.<port> > <chirpstack-host>.mps-raft: UDP, length 12
```

```
{"txpk":{"imme":false,"rfch":0,"powe":14,"ant":0,"brd":0,"tmst":4106230540,"freq":867.9,"modu":"LORA","datr":"SF10BW125","codr":"4/5","ipol":true,"size":18,"data":"YAV5GgaGBgAN2EmLUqW0KwmC"}}
12:21:42.957627IP <host>.<port> > <chirpstack-host>.mps-raft: UDP, length 12
```

These last two samples are for downlinks enqueued to devices in Mikrotik's coverage.

This one is from a downlink enqueued to a device in Milesight's coverage.

```
{"txpk":{"imme":true,"rfch":0,"powe":27,"ant":0,"brd":0,"freq":869.525,"modu":"LORA","datr":"SF12BW125","codr":"4/5","ipol":true,"size":18,"data":"YBK5hQaAAwAPDUgLCXr6j9jK"}}
12:24:40.925872 IP <host>.<port> > <chirpstack-host>.mps-raft: UDP, length 12

```

I'm not seeing on the tcpdump any tx_ack or pull_resp, neither from the Mikrotik nor the Milesight's gateways.
Thank you for your quick reply.

**Comment by brocaar** [comment]
Sorry, `PULL_RESP` is the packet type, the content is a `txpk`. For the `TX_ACK` the content is a `txpk_ack`:
https://github.com/Lora-net/packet_forwarder/blob/master/PROTOCOL.TXT#L408

Not seeing such package could mean two things:

* The gateway is unable to receive from or send data to your server (network related)
* The packet-forwarder on the gateway does not send the `TX_ACK` packet (software related)

In both cases this is an issue as ChirpStack uses this confirmation to know if the gateway was able to enqueue the packet for transmission. In case of a failure (or no reply), the downlink stays in the queue.

**Comment by joselunatic** [comment]
Thank you for your insights and suggestions. After delving deeper into the issue, I'm leaning towards the possibility that it's software-related, specifically with the packet forwarder in the Mikrotik gateways.

Here's a summary of my observations:

1. **Uplink Communication**: The Mikrotik gateways are successfully forwarding uplink messages to the network server. This aspect seems to be functioning correctly.
2. **Initial Downlink Success**: Interestingly, after a device successfully joins the network, the first downlink message is transmitted without any issues.
3. **Subsequent Downlink Failures**: However, following this initial success, the gateway seems unable to send any further downlink messages. The queue builds up but does not clear.
4. **Gateway Status**: Despite these downlink challenges, the Mikrotik gateways appear as 'Online' on the ChirpStack gateway page. They regularly send stats, indicating that their connection to the network server is stable.
5. **Geographical Replication of the Issue**: This problem isn't isolated to a single unit or location. I've observed the same behavior in two different Mikrotik units situated in distinct geographical areas.

These points lead me to believe that the issue might be rooted in the packet forwarder's handling of downlinks in the Mikrotik gateways. The consistency in uplink success, coupled with the repeated downlink failures post the initial transmission, suggests a specific bug affecting downlinks.

I would appreciate any further guidance or suggestions you might have on resolving this issue, particularly if there are known quirks or settings in the Mikrotik packet forwarder that could be contributing to this problem. Meanwhile, I'll raise an issue on Mikrotik's forum and will ask also for help there.

Thank you once again for your assistance.

**Comment by alyf80** [comment]
FYI, this (as well as #321) seems to be caused by a bug in the current implementation of the MikroTik forwarder: the token is byte-swapped in the `TX_ACK` packets.

```13:04:38.482895 IP 10.44.0.254.1700 > 10.44.0.31.56786: UDP, length 214
	0x0000:  4500 00f2 8c5d 4000 3f11 9929 0a2c 00fe  E....]@.?..).,..
	0x0010:  0a2c 001f 06a4 ddd2 00de 1664 025f 6803  .,.........d._h.
	0x0020:  7b22 7478 706b 223a 7b22 696d 6d65 223a  {"txpk":{"imme":
	0x0030:  6661 6c73 652c 2272 6663 6822 3a30 2c22  false,"rfch":0,"
	0x0040:  706f 7765 223a 3134 2c22 616e 7422 3a30  powe":14,"ant":0
	0x0050:  2c22 6272 6422 3a30 2c22 746d 7374 223a  ,"brd":0,"tmst":
	0x0060:  3131 3132 3631 3135 3837 2c22 6672 6571  1112611587,"freq
	0x0070:  223a 3836 382e 352c 226d 6f64 7522 3a22  ":868.5,"modu":"
	0x0080:  4c4f 5241 222c 2264 6174 7222 3a22 5346  LORA","datr":"SF
	0x0090:  3742 5731 3235 222c 2263 6f64 7222 3a22  7BW125","codr":"
	0x00a0:  342f 3522 2c22 6970 6f6c 223a 7472 7565  4/5","ipol":true
	0x00b0:  2c22 7369 7a65 223a 3333 2c22 6461 7461  ,"size":33,"data
	0x00c0:  223a 2249 4e4c 5757 2b4a 7168 692b 4247  ":"INLWW+Jqhi+BG
	0x00d0:  3948 3947 2b53 7131 3335 544b 3073 6970  9H9G+Sq135TK0sip
	0x00e0:  5775 5062 6664 444e 594f 6270 4978 7722  WuPbfdDNYObpIxw"
	0x00f0:  7d7d                                     }}

13:04:38.521822 IP 10.44.0.31.56786 > 10.44.0.254.1700: UDP, length 12
	0x0000:  4500 0028 e398 4000 4011 41b8 0a2c 001f  E..(..@.@.A..,..
	0x0010:  0a2c 00fe ddd2 06a4 0014 f502 0268 5f05  .,...........h_.
	0x0020:  3234 3336 2800 2200                      2436(.".
```
Note `5f68` in the `PULL_RESP` packet vs. `685f` in the `TX_ACK` response

Other ACK packets are unaffected

**Comment by Tatanka-77** [comment]
I've the same issue. The problem seems to be with mikrotik gateways running RouterOs > 7.11.
 Downgrade to 7.11 solve the problem (in my case...)

**Comment by brocaar** [comment]
I'm closing this issue as this seems to be an issue with the gateway firmware, not ChirpStack.

**Comment by alyf80** [comment]
For reference, this was fixed in RouterOS 7.13
